### PR TITLE
chore(hooks): remove callback support, closes #5228

### LINKED ISSUE #5228: Remove callback support for hooks

### PR DESCRIPTION
**Pull request by mickhansen**

### CODE REVIEW

File: lib/hooks.js [hunk 1]
```diff
@@ -114,46 +114,43 @@ const Hooks = {
...
     }
 
     if (!Array.isArray(hooks)) {
-      hooks = hooks === undefined ? [] : [hooks];
+      hooks = [hooks];
     }
 
-    // run hooks as sync functions if flagged as sync
+    // synchronous hooks
     if (hookTypes[hookType] && hookTypes[hookType].sync) {
```
Comment (felixfbecker) [inline lib/hooks.js]: Is this really needed? If the hook is sync, then `Promise.resolve()` will resolve immediately

Comment (mickhansen) [inline lib/hooks.js]: The code calling this is also sync.

Comment (felixfbecker) [inline lib/hooks.js]: Ah okay so the calling code relies on this returning instantly?

Comment (mickhansen) [inline lib/hooks.js]: Yes, i think the only caller currently is beforeDefine

File: test/integration/hooks/associations.test.js [hunk 2]
```diff
@@ -6,7 +6,8 @@ var chai = require('chai')
   , Support = require(__dirname + '/../support')
   , DataTypes = require(__dirname + '/../../../lib/data-types')
   , sinon = require('sinon')
-  , dialect = Support.getTestDialect();
+  , dialect = Support.getTestDialect()
+  , Promise = require('bluebird');
 
 describe(Support.getTestDialectTeaser('Hooks'), function() {
```
Comment (sushantdhiman) [inline test/integration/hooks/associations.test.js]: Almost all `function` expression here are missing space :laughing:, We should fix this formatting issue too

File: test/integration/hooks/associations.test.js [hunk 3]
```diff
@@ -62,14 +63,14 @@ describe(Support.getTestDialectTeaser('Hooks'), function() {
           , beforeHook = false
           , afterHook = false;
 
-          this.Tasks.beforeUpdate(function(task, options, fn) {
+          this.Tasks.beforeUpdate(function(task, options) {
             beforeHook = true;
-            fn();
+            return Promise.resolve();
```
Comment (sushantdhiman) [inline test/integration/hooks/associations.test.js]: What happens if a hook doesn't return anything ? Does it hangs forever ?

Comment (mickhansen) [inline test/integration/hooks/associations.test.js]: https://github.com/sequelize/sequelize/pull/6323/files/bcd8eb11dece81b885d573e801200669f26e402a#r71967651

File: test/integration/hooks/associations.test.js (hunk 2 above)
Comment (mickhansen) [inline test/integration/hooks/associations.test.js]: I agree, but style changes and functional changes should be kept seperate as much as possible so i'll leave it for another PR.

File: test/integration/hooks/associations.test.js (hunk 3 above)
Comment (sushantdhiman) [inline test/integration/hooks/associations.test.js]: Oh, [Like this :wink: ](https://www.google.co.in/search?client=opera&q=recursion&sourceid=opera&ie=UTF-8&oe=UTF-8) (may have some spelling mistakes)

Comment (mickhansen) [inline test/integration/hooks/associations.test.js]: Recursion is not relevant here? We wrap whatever response in Promise.resolve() so if it returns undefined it's just an undefined result in the resolved promise.

Comment (sushantdhiman) [inline test/integration/hooks/associations.test.js]: Your link generates recursion 😂, I just given my version of that. Without `Promise.resolve` this code should work, right?

Comment (mickhansen) [inline test/integration/hooks/associations.test.js]: Oh lol, damn.

Yes without Promise.resolve it works aswell, but using `Promise.resolve()` tests the intended async nature.

**Comment by sushantdhiman** [comment]
@mickhansen @janmeier LGTM, are there any other changes to make in this PR?

**Comment by janmeier** [comment]
Yep, LGTM as well

**Pull request merged**